driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/ReactiveContextProvider.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client;
+
+import com.mongodb.ContextProvider;
+import com.mongodb.RequestContext;
+import org.reactivestreams.Subscriber;
+
+/**
+ * A {@code ContextProvider} for reactive clients.
+ *
+ * @since 4.4
+ */
+public interface ReactiveContextProvider extends ContextProvider {
+    /**
+     * Get the request context from the subscriber.
+     *
+     * @param subscriber the subscriber for the operation
+     * @return the request context
+     */
+    RequestContext getContext(Subscriber<?> subscriber);
```

Comment:
> the parameter is (always!) a Subscriber

The parameter is currently a `Subscriber`. Is it actually inconceivable that if we were to use a different implementation of the reactive streams API, then this could have been a different object?

> we can use the type system to declare it

Regardless of the first part, in this case it's not a use but an abuse of the type system. If the type system were used correctly here, then implementations of the method would be able to use the object as a `Subscriber`, i.e., call the methods declared by this interface, but they absolutely must not do so.

Is there is a good reason to abuse the type system here? Regardless of whether the parameter type is `Subscriber` or `Object`, the implementation needs to have arcane knowledge about possible runtime types that that this object may be of, whether they bear a context, and how a context can be extracted. The knowledge about the object being a `Subscriber` helps to narrow down the set of runtime types which still comes not from our API, but via other information channels, and is used when writing the code. I.e., it has no use at both compile- and run time, which means that there is no way to justify the abuse of the type system here.

Reply:
> The parameter is currently a Subscriber. Is it actually inconceivable that if we were to use a different implementation of the reactive streams API, then this could have been a different object?

Yes, that wouldn't happen.  Our API is defined not in terms of Project Reactor but Reactive Streams, which is where Subscriber is defined.  All we are saying here is that the object that the application passes to `org.reactivestreams.Publisher#subscribe is what we're passing to this method.  And since subscribe takes an instance of Subscriber, that's the type we use here as well.

It does have some use at compile time, since any attempt to cast this object to a type that is not an instance-of Subscriber will be a compile time error.  But if it where Object, then 

`String whatever = (String) subscriber;`

would compile.

Probably best to discuss in person if you want to debate this further. Not sure we're getting anywhere.

Reply:
Indeed, the point about allowing implementation to cast only to subtypes of `Subscriber` is valid, and I did not think about it. With your argument in mind, the decision does not look black and white to me anymore. Thanks! I agree that having the API as it is now can be justified.